Updated to include more Message types

davidpiesse opened the pull request:
Added more message types (basic text, button, generic, attachment)
Changed API version to 2.7
Updated README to reflect changes
Tweaked directory structure to include Components (Button, Card)
Moved around Exceptions into better organised classes (mostly from Send Error to Message Error)
Attempted to keep to style guide :)

davidpiesse committed:
Changed way Facebook Message is built.
Changed API to v2.7
Added 2 extra exceptions
Separated out Components and Enums
Added alternative templates including Button, Generic, Text and Attachment

davidpiesse committed:
Altered way exceptions are organised into Notification, Message, and each component

davidpiesse committed:
Updated Readme File to reflect changes

davidpiesse committed:
Removed duplicate Notification Type Class

irazasyed commented on README.md:
```diff
@@ -63,11 +63,13 @@ Next we need to add this token to our Laravel configurations. Create a new Faceb
 Let's take an invoice-paid-notification as an example.
 You can now use the Facebook channel in your `via()` method, inside the InvoicePaid class. The `to($userId)` method defines the Facebook user, you want to send the notification to.
 
+Based on the details you add (text, attachments etc.) will determine automatically the type of message to be sent. For example if you only add `text()` then it will be a basic message; using `attach()` will turn this into a attachment message. Having `buttons` or `cards` will cahneg this to the `Button Template` and `Generic Template` respectivily
```
There's a typo:

> Having `buttons` or `cards` will **cahneg** this to the ...

Fix: **change**

irazasyed commented on README.md:
```diff
@@ -121,25 +123,36 @@ public function routeNotificationForFacebook()
 - `title('')`: (string) Button Title.
 - `data('')`: (string) Button Data - It can be a web url, postback data or a formated phone number.
-- `type('')`: (string) Button Type - `web_url`, `postback` or `phone_number`.
+- `type('')`: (string) Button Type - `web_url`, `postback` or `phone_number`. Use `ButtonType` enumerator for garunteeing valid values
```
Typo:

> Use `ButtonType` enumerator for **garunteeing** valid values ...

Fix: **guaranteeing**

irazasyed commented on README.md:
```diff
@@ -121,25 +123,36 @@ public function routeNotificationForFacebook()
+- `title('')`: (string) Card Title.
+- `subtitle('')`: (string) Card Subtitle.
+- `url('')`: (string) Card Item Url.
+- `image('')`: (string) Card Image Url. Image ratio hould be 1.91:1
```
Typo:

> Image ratio **hould** be 1.91:1

Fix: **should**

irazasyed commented on README.md:
```diff
@@ -121,25 +123,36 @@ public function routeNotificationForFacebook()
 ## Credits
 
 - [Syed Irfaq R.](https://github.com/irazasyed)
+- [David Piesse](https://github.com/davidpiesse)
```
Contributions from individual contributors are credited in [CHANGELOG](https://github.com/laravel-notification-channels/facebook/blob/master/CHANGELOG.md) and in release notes. We can't really credit each contributor individually in README.

So you can update the [CHANGELOG](https://github.com/laravel-notification-channels/facebook/blob/master/CHANGELOG.md) file with all the changes you've made and either credit yourself or will do it before release.

irazasyed commented on src/Components/Card.php:
```diff
@@ -0,0 +1,157 @@
+        $this->buttons = $buttons;
+        return $this;
+    }
+
```
This `buttons` method seems to be duplicated in another file too. Hence, It's better we move this to a trait instead. Please let me know if you can do it or I'll do it after merging.

davidpiesse replied:
If you can make it a trait that would be super helpful.
I don't know of anywhere else off the top of my head we would need it in fb
messaging

On 20 Aug 2016 23:49, "Syed Irfaq R. ⚡️" notifications@github.com wrote:

> In src/Components/Card.php
> https://github.com/laravel-notification-channels/facebook/pull/10#discussion_r75585974
> :
> 
> > -     \* Add up to 3 call to action buttons.
> > -     *
> > -     \* @param array $buttons
> > -     *
> > -     \* @return $this
> > -     \* @throws CouldNotSendNotification
> > -     */
> > -    public function buttons(array $buttons = [])
> > -    {
> > -        if (count($buttons) > 3) {
> > -            throw CouldNotCreateMessage::messageButtonsLimitExceeded();
> > -        }
> > -        $this->buttons = $buttons;
> > -        return $this;
> > -    }
> >   +
> 
> This buttons method seems to be duplicated in another file. Hence, It's
> better we move this to a trait instead. Please let me know if you can do it
> or I'll do it after merging.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/laravel-notification-channels/facebook/pull/10/files/f44a70e62797eeef04e9355223738d90b297d1e9#r75585974,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAw3ikVG5EAf80cewI4nDtWfqkXOBQyHks5qh4STgaJpZM4JpLCo
> .

irazasyed replied:
I'll add it once I merge this then :)

irazasyed commented on src/FacebookMessage.php:
```diff
@@ -58,13 +90,43 @@ public function to($recipient)
+    public function attach($attachment_type, $url)
+    {
+        if (in_array($attachment_type, [AttachmentType::FILE, AttachmentType::IMAGE,
+            AttachmentType::VIDEO, AttachmentType::AUDIO]))
```
Can we move that array above the method like so? :

``` php
$attachment_types = [AttachmentType::FILE, AttachmentType::IMAGE, AttachmentType::VIDEO, AttachmentType::AUDIO];

if (in_array($attachment_type, $attachment_types))
```

It's for readability purposes.

irazasyed commented:
Thanks for opening a PR. I appreciate it.

I went through the code and commented where there needs to be a change done, please take a look and apply the changes when you get a chance.

Overall everything seems good but the code style is not really consistent and there are other minor issues which I'll fix myself with a patch once this gets merged.

Thanks!

davidpiesse committed:
Added ButtonsTrait for Facebook Message and also Card.
Fixed typo's
Reformatted slightly for StyleCI

davidpiesse commented:
Now added the ButtonsTrait, fixed found typo's and added some more examples to the README.md.
Implemented code style change and fixed what I could.
Let me know if there are any more changes you want to happen before a merge :)

irazasyed commented on README.md:
```diff
@@ -63,7 +63,7 @@ Next we need to add this token to our Laravel configurations. Create a new Faceb
 You can now use the Facebook channel in your `via()` method, inside the InvoicePaid class. The `to($userId)` method defines the Facebook user, you want to send the notification to.
 
-Based on the details you add (text, attachments etc.) will determine automatically the type of message to be sent. For example if you only add `text()` then it will be a basic message; using `attach()` will turn this into a attachment message. Having `buttons` or `cards` will cahneg this to the `Button Template` and `Generic Template` respectivily
+Based on the details you add (text, attachments etc.) will determine automatically the type of message to be sent. For example if you only add `text()` then it will be a basic message; using `attach()` will turn this into a attachment message. Having `buttons` or `cards` will chaneg this to the `Button Template` and `Generic Template` respectivily
```
**change** <-- Still a type in your text :)

irazasyed commented on README.md:
````diff
@@ -101,6 +101,39 @@ The notification will be sent from your Facebook page, whose page token you have
+->to($this->user->fb_messenger_id);
+```
+##### Attachment Message
+Send a file attachment to a user (Example is sending a pdf invoice)
````
Not sure if the below attachment is of PDF or an image? Because it says `AttachmentType::IMAGE` with a URL. It doesn't seem PDF.

irazasyed commented on README.md:
````diff
@@ -101,6 +101,39 @@ The notification will be sent from your Facebook page, whose page token you have
+```
+
+##### Generic (Card Carousel) Message
+Send a set fo cards / items to a user displayed in a corousel (Example is sending a set of links). Note you can also add up to three buttons per card
````
Send a set **of** cards / items to a user displayed in a **carousel**

irazasyed commented:
Just some more typo fixes and we should be all set to merge this! Please see if you can fix them.

davidpiesse commented:
Yep!!
Will do today!!

Thanks for the all the reviewing :)
D

On 21 Aug 2016 16:41, "Syed Irfaq R. ⚡️" notifications@github.com wrote:

> Just a some more typos and we should be all set to merge this! Please see
> if you can fix them.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/laravel-notification-channels/facebook/pull/10#issuecomment-241264727,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAw3ipDBkoWP-2UFLk4LJkWShOR6-n1hks5qiHGtgaJpZM4JpLCo
> .

davidpiesse committed:
Typo fixes

Changed typos as found by Syed.

irazasyed commented:
Merged. Thank you for your contribution 👍

davidpiesse commented:
No problems!
I will endeavour too add more to it.
I am also going to refactor my own Facebook Messaging repo to incorporate
some of the things learnt here :)

D

On 22 August 2016 at 17:13, Syed Irfaq R. ⚡️ notifications@github.com
wrote:

> Merged. Thank you for your contribution 👍
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/laravel-notification-channels/facebook/pull/10#issuecomment-241465015,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAw3ioXKZV0LpUsCaucTIjMQpozNbxvHks5qicqwgaJpZM4JpLCo
> .
